src/basilisp/core.lpy

```diff
@@ -7085,3 +7085,122 @@
+                                         nil))]
+                    [(str fname)
+                     `(fn [& ~vargs]
+                        ;; create a case form to switch to the right
+                        ;; body according to the number of arguments
+                        ;; passed to the fn.
+                        ;;
+                        ;; (case (count fn-args)
+                        ;;   n1 (do body1)
+                        ;;   n2 (do body2)
+                        ;;   (if-not (variadic? fn-args)
+                        ;;     (throw exception)
+                        ;;     (do variadic-body))
+                        ~(apply list 'case (list 'count vargs)
+                                (mapcat (fn [args-and-body]
+                                          (let [args (first args-and-body)
+                                                body (rest args-and-body)]
+                                            (if (and (= args-and-body fargs-last)
+                                                     variadic-pos)
+                                              `((if (<= (count ~vargs) ~variadic-pos)
+                                                  (throw (ex-info "Arguments passed to function does not match the variadic fn specification"
+                                                                  {:args-spec-expected (quote ~(first fargs-last))
+                                                                   :args-provided (drop 1 ~vargs)}))
+                                                  (let [~(into ['this] args) ~vargs]
+                                                    (do ~@body))))
+                                              [(inc (count args)) (list 'let [(into ['this] args) vargs]
+                                                                        `(do ~@body))])))
+                                        fargs)))])
+
+
+                  ;; Implementation of variadic functions for Basilisp
+                  ;; interfaces necessitates the inclusion of
+                  ;; additional arity method signatures, such as
+                  ;; method-name_arityN, to be registered for proper
+                  ;; functionality, even if their implementations are
+                  ;; nil.
+                  fname-fndummy-pairs
+                  (for [spec (rest f)]
+                    (let [fargs (first spec)
+                          argc (count fargs)
+                          variadic? (try
+                                      (.index fargs '&)
+                                      (catch python/ValueError _
+                                        nil))
+                          arity (if variadic?
+                                  "_rest"
+                                  (if (= argc 0) 0 (inc argc)))]
+                      [(str "_" (first f) "_arity" arity) nil]))]
+              (into [fname-fn-pair] fname-fndummy-pairs))
```

Comment:
I'm curious why you implemented these as dummy methods rather than just implementing each arity directly? The generated Python for multi-arity functions already does this. It uses a pure Python version of the case strategy you use above but ultimately calls a function rather than putting the code inline.

I'm wondering if you wouldn't get a lot of this logic for free if you just rewrote the multi-arity methods directly as multi-arity fns with `this` prepended to the argument vectors. 

My concern is basically that we're going to have dummy methods to appease the ABC, but those methods do exist and can be called but won't work like they would for any other implementation.

Reply:
It is still not clear how the multi-arity fns are implement internally and as such is difficult for me to understanding why both the dispatch (the 'case' fn) and the individual arity fns are needed.

I've provided an example at the end of this command how the macro expansion is done with my latest update

With the latest update, I populate all `multi`, `_multi_arity0`, `_multi_arity2` and `_multi_arity_rest` fn defintion, where previously I was only providing the first and the rest were `nil`. Though there is no calling one another here, it is more like providing a fn definition twice, one for `multi` and another for the corresponding arity fn (e.g. `_multi_arity2`). Unfortunately I can't  `(apply ._multi_arity2 (first args) (rest args))` from within the `multi` fn so as to call the arity fn, since `apply`'ing to a method doesn't seem to work.

I've also tried to set the `multi` method to a multiarity fn, but it  still complains that `_multi_arity0` etc are not defined.

Perhaps you can guide me with an example how you think the `type` fn dictionary should be better structured, if there is a better alternative?

Thanks

example
``` clojure
(proxy [ITestProxyMultiVariadic] []
  (multi
    ([] 0)
    ([atm]
     (swap! atm inc)
     @atm)
    ([one two three & more] [one two three more])))
```

macro expands to
``` clojure
((python/type
  "proxy--ITestProxyMultiVariadic_14748"
  (python/tuple [ITestProxyMultiVariadic])
  (basilisp.core/lisp->py
   {"multi"
    (basilisp.core/fn
      [& args_14749]
      (basilisp.core/case
          (basilisp.core/count args_14749)
          1
          (basilisp.core/apply (basilisp.core/fn [this] 0) args_14749)
          2
          (basilisp.core/apply
           (basilisp.core/fn
             [this atm]
             (swap! atm inc)
             (basilisp.core/deref atm))
           args_14749)
          (if
              (basilisp.core/<= (basilisp.core/count args_14749) 3)
              (throw
               (basilisp.core/ex-info
                "Arguments passed to function does not match the variadic fn specification"
                {:args-spec-expected '[one two three & more],
                 :args-provided (basilisp.core/drop 1 args_14749)}))
              (basilisp.core/apply
               (basilisp.core/fn
                 [this one two three & more]
                 [one two three more])
               args_14749)))),
    "_multi_arity0" (basilisp.core/fn [this] 0),
    "_multi_arity2"
    (basilisp.core/fn
      [this atm]
      (swap! atm inc)
      (basilisp.core/deref atm)),
    "_multi_arity_rest"
    (basilisp.core/fn
      [this one two three & more]
      [one two three more])})))
```